exchanges/exchanges.go

```diff
@@ -497,6 +523,98 @@ func (xc *CommonExchange) state() *ExchangeState {
 	return xc.currentState
 }
 
+// WebsocketProcessor is a callback for new websocket messages from the server.
+type WebsocketProcessor func([]byte)
+
+// Only the fields are protected for these, so okay to grab them like this.
```

Comment:
Gorilla's read and write methods need concurrency control. The `websocketFeed` itself is safe for concurrent use, but the `commonExchange` field will be updated during a reconnection. 

For `websocketFeed`, writing is protected by the `(socketClient).writeMtx`, and the read is good because `commonExchange` has a single read loop assigned to each connection. 

I was trying to say that once you have the `websocketFeed`, which as you said is a pointer underneath, it is safe to use concurrently. The `commonExchange.ws*` fields need protection, though. 

No need to match request with response for the websocket feeds, but there is a sequence ID I should probably check to make sure orderbook updates are processed in order.